cansav09/scRNA-seq_tutorial (#19)
AlexsLemonade/training-modules: scRNA-seq/pre-processing_scRNA-seq.md

being one of two groups, based on their capture methods and quantitative nature.
Depending on how the cells are sorted and what technology is used, the pre-processing steps are a bit different and the biases to look out for in post-processing also vary. 

For a more extensive run through on single-cell technologies, 
Kiselev at al have very [good tutorial for scRNA-seq](https://hemberg-lab.github.io/scRNA.seq.course/introduction-to-single-cell-rna-seq.html#experimental-methods). 
  
### 1) Non-Tag-Based scRNA-seq
*Examples: Smart-seq2*  
Cells are physically separated generally into individual wells in a plate and 
often also sorted by other means (eg. Fluorescence Activated Cell Sorting). 
Each cell is then sequenced individual and has it's own fastq file (this will be two fastq files if this is paired-end sequencing.) 
The data pre-processing steps for these types of scRNA-seq data is/can be more similar to 
bulk rna-seq methods.

#### Pros:  
- Can be paired end sequencing which has less risk for 3' bias.  
- More complete coverage of transcripts, which may be better for transcript 
discovery purposes.   
  
#### Cons:  
- Is not very efficient (generally 96 cells per plate)  

> - Takes much longer to run.  

Author: Looked for a number in the literature, didn't find anything great to describe it. Put a general  timeline statement to try to convey this point.

  
### 2) Tag-based scRNA-seq
*Examples: 10X Genomics, Drop-seq*  
Cells are separated by emulsion/droplets, and individual cells are given barcodes. 
Everything is then sequenced. 
These types of methods, because they are newer, are more likely to have Unique 
Molecular Identifiers (UMIs) which allow you to better control for PCR amplification errors and biases.
Individual samples have two fastq files: one for the cell barcodes
and another with the individual reads. 

#### Pros:  
- Can run potentially millions of cells at once.   
- A lot quicker computing wise.  
- Won't take up all your computer's storage.  
- A lot cheaper  
  
#### Cons:  
- Sequencing is not bidirectional so data will likely have more intense 3' bias.  
- Coverage of these technologies generally is not as deep.  

### Processing scRNA-seq data

#### Step 0: Obtaining the data
10X Genomics has plenty of [datasets](https://support.10xgenomics.com/single-cell-gene-expression/datasets) 
available to play around with. 
For this example, we will use the [1k pbmc](https://support.10xgenomics.com/single-cell-gene-expression/datasets/3.0.0/pbmc_1k_v2) 
(peripheral blood mononuclear cells). 

> We will use one of these samples' subsetted fastq file of this as an example.  

Author: Flashdrives make sense I think.

  
*Note*: depending on the state of the data you are working with, ie. if you have 
a `.bcl` file, you will need to use `CellRanger` with their `mkfastq` command to 
make fastq files from this.
However, because most publicly available data is in fastq format and any
core you might be working with will also likely provide you with fastq files, 
we are starting from this point.  

#### Step 1: Set up your directories
As described in a previous module, we'll make some directories from command line
for us to store these data in. 
``` 
mkdir data
mkdir ref_files
mkdir results
mkdir data/alevin_output
```

> #### Step 2: Index the human transcriptome with Salmon

Author: I think we mentioned this in person: plan is to make the index. This index will like have to be slightly different than the bulk RNA-seq index,  because it probably should have a smaller k.

**CCDL 2019**

#### In this section, we will be running through the basics of pre-processing
#### single-cell RNA-seq data.

For the purposes of this tutorial, we'll summarize single-cell technologies as
being one of two groups, based on their capture methods and quantitative nature.
Depending on how the cells are sorted and what technology is used, the pre-processing steps are a bit different and the biases to look out for in post-processing also vary.

For a more extensive run through on single-cell technologies,
Kiselev at al have very [good tutorial for scRNA-seq](https://hemberg-lab.github.io/scRNA.seq.course/introduction-to-single-cell-rna-seq.html#experimental-methods).

### 1) Non-Tag-Based scRNA-seq  
*Example:* Smart-seq2 [(Picelli et al, 2014)](https://www.nature.com/articles/nprot.2014.006)   
Cells are physically separated generally into individual wells in a plate and
often also sorted by other means (eg. Fluorescence Activated Cell Sorting).
Each cell is then sequenced individual and has it's own fastq file (this will be two fastq files if this is paired-end sequencing.)
The data pre-processing steps for these types of scRNA-seq data is/can be more similar to

> bulk rna-seq methods.

Reviewer: Doesn't look like this change has been made as I'm viewing it on Github.
```suggestion
bulk RNA-seq methods.
```

Author: Yeah, I remember having already made this change. But maybe I'm thinking of a different module? Oh well, I've remade the changes here.

#### Pros:  
- Can be paired end sequencing which has less risk for 3' bias.  
- More complete coverage of transcripts, which may be better for transcript
discovery purposes.   

#### Cons:  

> - Is not very efficient (generally 96 cells per plate)  

Reviewer: ```suggestion
- Is not very efficient (generally 96 cells per plate).
```

> - Takes much longer to run (Days/weeks depending on sample size) .

Reviewer: ```suggestion
- Takes much longer to run (Days/weeks depending on sample size).
```

- Is a lot more expensive.  

### 2) Tag-based scRNA-seq  
*Example:* 10X Genomics [(Zheng et al, 2017)](https://www.ncbi.nlm.nih.gov/pubmed/28091601)  
Cells are separated by emulsion/droplets, and individual cells are given barcodes.
Everything is then sequenced.
These types of methods, because they are newer, are more likely to have
[Unique Molecular Identifiers (UMIs)](http://www.nature.com/doifinder/10.1038/nmeth.2772)
which allow you to better control for PCR amplification errors and biases.
Individual samples have two fastq files: one for the cell barcodes
and another with the individual reads.

#### Pros:  
- Can run potentially millions of cells at once.   
- A lot quicker computing wise.  
- Won't take up all your computer's storage.  

> - A lot cheaper  

Reviewer: ```suggestion
- A lot cheaper.
```
if you're ending everything else with a period

#### Cons:  
- Sequencing is not bidirectional so data will likely have more intense 3' bias.  
- Coverage of these technologies generally is not as deep.  

> *More sources on the comparisons and explanations of these technologies: *   

Reviewer: I think you intended for this to be bold
```suggestion
*More sources on the comparisons and explanations of these technologies:*   
```

available to play around with.
For this example, we will use the [1k pbmc](https://support.10xgenomics.com/single-cell-gene-expression/datasets/3.0.0/pbmc_1k_v2)
(peripheral blood mononuclear cells).
We will process a fastq file of this as an example.  
For practical purposes as far as time constraints, we have subset this fastq file
and have it ready for you to process but with only the first 3 million reads.

*Note*: depending on the state of the data you are working with, ie. if you have
a `.bcl` file, you will need to use `CellRanger` with their `mkfastq` command to
make fastq files from this.
However, because most publicly available data is in fastq format and any
core you might be working with will also likely provide you with fastq files,
we are starting from this point.  

### Step 1: Set up your output directory
As described in a previous module, we'll make a directory from command line
for us to store our output data in.
```
mkdir alevin_output  
```

### Step 2: Index the human transcriptome with Salmon
Before you can quantify with Salmon and
[Alevin](https://www.biorxiv.org/content/10.1101/335000v2), we need a transcriptome
to be indexed.

> You can use the same trancriptome index as was used for bulk-rna-seq, however,

Reviewer: ```suggestion
You can use the same transcriptome index as was used for bulk RNA-seq, however,
```

> due to the smaller pieces and amounts of single cell RNA-seq as opposed to bulk,

Reviewer: It's a matter of read length if we're keeping it simple and I am inclined to keep it simple.

you may want to build the index with a smaller `-k`.

> In this instance, we used a `-k` of 23 using the ensemble transcriptome.

Reviewer: ```suggestion
In this instance, we used a `-k` of 23 using the Ensembl transcriptome.
```

In the interest of time, we have already run the command below and have the index
built for you.
But for your own reference, here is how you'd do it yourself:
```
salmon --threads=16 --no-version-check index \
  -t Homo_sapiens.GRCh38.cdna.all.fa.gz \
  -i human_index \
  -k 23
```

### Step 3: For each sample, run [Alevin](https://www.biorxiv.org/content/10.1101/335000v2)
for quantification
From the command line, running Alevin is not too much different from running

> Salmon for bulk rna-seq. You'll recognize a lot of these options as the same.

Reviewer: ```suggestion
Salmon for bulk RNA-seq. You'll recognize a lot of these options as the same.
```

In this instance, files with `R1` contain the barcodes for the cells as well as
the Unique Molecular Identifiers while `R2` files contain the full reads for that sample.  

#### `-l`
The ISR library type is what is recommended for single cell data quant.

#### `--chromium`

> Because we are using 10X chromium data, we have to use this flag. However,

Reviewer: ```suggestion
Because we are using 10X Chromium data, we have to use this flag. However,
```
maybe?

Drop-seq data is also supported, and in this case you would use a `--dropseq`
flag instead of this.

#### `--tgMap`
This is needed to supply a trasncript to gene key that Alevin will use to
quantify the genes. For our example, we've premade the file `genes_2_tx.tsv` from

> the ensembl transcriptome that we indexed above. The file has to be a .tsv file.

Reviewer: ```suggestion
the Ensembl transcriptome that we generated with `salmon index` above. The file has to be a .tsv file.
```

You say csv (no period below), I think the preferred style for these would be TSV and CSV.

#### `--dumpCsvCounts`
Using this option will make the counts in a csv file and make it easier for us to
import into R later.

#### `--dumpFeatures`
This option will print out information that we will need for quality checks
later on, including files with information on the UMI's and cell barcodes ("CB").  

#### Running Salmon Alevin

> Coy and paste this in your command line to run Alevin quantification

Reviewer: ```suggestion
Copy and paste this in your command line to run Alevin quantification
```
